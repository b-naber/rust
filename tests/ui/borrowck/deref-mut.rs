use std::cell::RefMut;

struct Foo {
    a: u32,
}

fn func1(y: RefMut<'_, Foo>) {
    let shared_ref = &y;
    *y = Foo { a: 3 };
    //~^ ERROR cannot borrow `y` as mutable because it is also borrowed as immutable
    takes_shared_foo_ref(shared_ref);
}

fn func2(y: RefMut<'_, Foo>) {
    let shared_ref = &y.a;
    *y = Foo { a: 3 };
    //~^ ERROR cannot borrow `y` as mutable because it is also borrowed as immutable
    takes_shared_u32_ref(shared_ref);
}

fn func3(y: RefMut<'_, Foo>) {
    let shared_ref = &y.a;
    y.a = 3;
    //~^ ERROR cannot borrow `y` as mutable because it is also borrowed as immutable
    takes_shared_u32_ref(shared_ref);
}

fn takes_shared_foo_ref<'a>(x: &'a RefMut<'_, Foo>) {}
fn takes_shared_u32_ref<'a>(x: &'a u32) {}

fn main() {}
