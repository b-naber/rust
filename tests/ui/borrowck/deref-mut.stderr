error[E0502]: cannot borrow `y` as mutable because it is also borrowed as immutable
  --> $DIR/deref-mut.rs:9:6
   |
LL |     let shared_ref = &y;
   |                      -- immutable borrow occurs here
LL |     *y = Foo { a: 3 };
   |      ^ mutable borrow occurs here
LL |
LL |     takes_shared_foo_ref(shared_ref);
   |                          ---------- immutable borrow later used here

error[E0502]: cannot borrow `y` as mutable because it is also borrowed as immutable
  --> $DIR/deref-mut.rs:16:6
   |
LL |     let shared_ref = &y.a;
   |                       - immutable borrow occurs here
LL |     *y = Foo { a: 3 };
   |      ^ mutable borrow occurs here
LL |
LL |     takes_shared_u32_ref(shared_ref);
   |                          ---------- immutable borrow later used here

error[E0502]: cannot borrow `y` as mutable because it is also borrowed as immutable
  --> $DIR/deref-mut.rs:23:5
   |
LL |     let shared_ref = &y.a;
   |                       - immutable borrow occurs here
LL |     y.a = 3;
   |     ^ mutable borrow occurs here
LL |
LL |     takes_shared_u32_ref(shared_ref);
   |                          ---------- immutable borrow later used here

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0502`.
